### PR TITLE
Add setting: always add "Cited on pages" text to JStyles.

### PR DESCRIPTION
**Pull request by heyitsdross**
Closes https://github.com/JabRef/jabref/issues/11691
### TODO (for me):
[] Update user documentation, open documentation PR.
[] Review developer documentation to see if update necessary. (I think it's unlikely).

### To Test:
1. Toggle the Open- / LibreOffice view under `View > OpenOffice/LibreOffice` (or `Alt + 0`)
2. Click on "Select Style" in the OpenOffice menu which appears (assumedly on the left side of the window).
3. Choose any JStyle.
4. In the same area as "Select Style", click on "Settings"
5. **Expected**: The option "Automatically add "Cited on pages..." at beginning of citation" should appear unchecked.
6. Toggle the setting exactly once by clicking.
7. Fully close and restart the application.
8. Repeat steps 1 (if the OpenOffice view doesn't reopen) and 4.
9. **Expected**: The toggled option should still be checked.
10. Optionally repeat steps 6, then 1 and 4 again. The now twice toggled setting should again still be preserved (aka it should be unchecked).

### Screenshot Preview of feature:
11691-pr-ui-screenshot

### Mandatory checks

- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

**Review by subhramit — CHANGES_REQUESTED** [review]
Hey, thanks for taking out time to work on this.
You have gotten the preferences part right, as well as the UI (except for the change listener - check review comments).
Regarding the backend - the following variable is still always set to false, which still renders the functionality useless:
https://github.com/JabRef/jabref/blob/3df3b4fd427b27e9ac8ef603c10996d7118f1f9a/src/main/java/org/jabref/gui/openoffice/OOBibBase.java#L72

Please update it's state according to the user's selection. Should be easy passing it to `OOBibBase`, see how the other settings are evaluated there.

I have also left some general comments. I also suggest going through the issue once again for more clarity.

**Comment by subhramit** [comment]
@InAnYan Hey Ruslan, although you've done it in https://github.com/JabRef/jabref/pull/11727, I think `aiService` is not required here as well?
https://github.com/JabRef/jabref/blob/3df3b4fd427b27e9ac8ef603c10996d7118f1f9a/src/main/java/org/jabref/gui/openoffice/OpenOfficePanel.java#L98

File: src/main/java/org/jabref/gui/openoffice/OpenOfficePanel.java [hunk 1]
```diff
@@ -618,12 +618,16 @@ private boolean checkThatEntriesHaveKeys(List<BibEntry> entries) {
     }
 
     private ContextMenu createSettingsPopup() {
-        OpenOfficePreferences openOfficePreferences = preferencesService.getOpenOfficePreferences();
-
```
[the remark this answers is not in the document]
Comment (heyitsdross) [reply src/main/java/org/jabref/gui/openoffice/OpenOfficePanel.java:622]: Personal practice is to initialize as close to first usage as possible to increase refactorability, but if it's project standard to initialize at the beginning, then I will gladly undo.

File: src/main/java/org/jabref/gui/openoffice/OpenOfficePanel.java [hunk 2]
```diff
@@ -657,4 +661,15 @@ private ContextMenu createSettingsPopup() {
...
+
+    private void addAlwaysCitedOnPagesTextOption(ContextMenu contextMenu) {
+        OpenOfficePreferences openOfficePreferences = preferencesService.getOpenOfficePreferences();
+
+        CheckMenuItem alwaysAddCitedOnPagesText = new CheckMenuItem(Localization.lang("Automatically add \"Cited on pages...\" at beginning of citation"));
+
+        alwaysAddCitedOnPagesText.selectedProperty().set(openOfficePreferences.getAlwaysAddCitedOnPages());
+        alwaysAddCitedOnPagesText.setOnAction(e -> openOfficePreferences.setAlwaysAddCitedOnPages(alwaysAddCitedOnPagesText.isSelected()));
+
+        contextMenu.getItems().add(alwaysAddCitedOnPagesText);
```
[the remark this answers is not in the document]
Comment (heyitsdross) [reply src/main/java/org/jabref/gui/openoffice/OpenOfficePanel.java:673]: I'm really embarrassed I forgot about the CStyles, will add the removal mechanism and move the item below the "old" option as requested.

File: src/main/java/org/jabref/gui/openoffice/OpenOfficePanel.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (subhramit) [reply src/main/java/org/jabref/gui/openoffice/OpenOfficePanel.java:622]: Yes, there is no general hard-and-fast best practice regarding this. Beginning-of-scope-initialization is preferred when a dependency is used throughout the scope multiple times. Near-first-use-initialization is used if the usage is confined, and refactorability is indeed the argument.

I would have requested you to change back, but since moving it to the class level (my comment below on refactoring) will anyway do it at a more global level, this line will just be removed here.

File: src/main/java/org/jabref/gui/openoffice/OpenOfficePanel.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (subhramit) [reply src/main/java/org/jabref/gui/openoffice/OpenOfficePanel.java:673]: > I'm really embarrassed I forgot about the CSL Styles, will add the removal mechanism and move the item below the "old" option as requested.

Please be comfortable, you are new to the codebase, so that's very normal!

File: src/main/java/org/jabref/gui/openoffice/OOBibBase.java [hunk 3]
```diff
@@ -68,22 +69,21 @@ public class OOBibBase {
...
+    public OOBibBase(Path loPath, DialogService dialogService, OpenOfficePreferences openOfficePreferences)
             throws
             BootstrapException,
             CreationException {
 
         this.dialogService = dialogService;
         this.connection = new OOBibBaseConnect(loPath, dialogService);
 
-        this.alwaysAddCitedOnPages = false;
+        this.alwaysAddCitedOnPages = openOfficePreferences.getAlwaysAddCitedOnPages();
```
Comment (heyitsdross) [inline src/main/java/org/jabref/gui/openoffice/OOBibBase.java:86]: I wasn't sure here if I should inject the boolean value itself, the `OpenOfficePreferences` or the `PreferencesService`, I opted to stay closer to the real object while maintaining a layer of abstraction.

Let me know if this was appropriate for our standards or not.

Comment (Siedlerchr) [reply src/main/java/org/jabref/gui/openoffice/OOBibBase.java:86]: Yes this is fine. Especially since in the background the preferencs are observables you make sure that you are always getting the latest value

**Comment by subhramit** [comment]
Great work, Austin!
I took the opportunity to make some minor changes, you can check my commit (https://github.com/JabRef/jabref/pull/11732/commits/21c7de17e411df88b2dca2e37046bf1c1d55827a). Mostly changed the dialog string, shifted the changelog entry down (we follow entries to be in chronological and sections in reverse chronological order) and some minor nitpicks.

Only one last thing to fix - you have accidentally committed submodules (see `abbrv.jabref.org` and `csl-styles` in changed files). To fix, follow:
```
$ git merge origin/main
$ git checkout main -- src/main/resources/csl-styles
$ ... git commit ... 
$ git push
```
And similarly for `abbrv.jabref.org`.

To avoid this: avoid using `git add .` to stage from CLI. Preferably use a GUI-based git manager, such as the one built in IntelliJ or open `git gui` from the command line. Even if you accidentally stage them, don't commit all files, selectively commit the files you touched using the GUI based tool, and push.

Context: JabRef needs submodules such as CSL style files for some of its respective features. When setting up a local development environment, these are set up once using `--recurse-submodules` (you may have noticed). These submodules, in the main branch, are automatically periodically updated but not fetched into local again when you pull, as they are set to be ignored in `.gitignore` (this is to avoid merge conflicts). So when `remote:main` has updated submodules, and your local has the old ones, when you stash, these changes are noticed.

File: src/main/java/org/jabref/gui/openoffice/OpenOfficePanel.java [hunk 4]
```diff
@@ -657,4 +661,15 @@ private ContextMenu createSettingsPopup() {
 
         return contextMenu;
     }
+
+    private void addAlwaysCitedOnPagesTextOption(ContextMenu contextMenu) {
+        OpenOfficePreferences openOfficePreferences = preferencesService.getOpenOfficePreferences();
+
+        CheckMenuItem alwaysAddCitedOnPagesText = new CheckMenuItem(Localization.lang("Automatically add \"Cited on pages...\" at beginning of citation"));
```
[the remark this answers is not in the document]
Comment (ThiloteE) [reply src/main/java/org/jabref/gui/openoffice/OpenOfficePanel.java]: Good job. It definitely should be at the end of the citation. "[backref](https://ctan.org/pkg/backref?lang=en)", which is part of the hyperref package offers the "cited on pages" feature for LaTeX documents, so that could serve as an inspiration. I do not remember exactly, but I believe I made some minor adjustments to the default backref settings to always start in a new line, which I think looks slick :-)

Here an example from a paper I wrote in LaTeX that includes the backref feature:
image

**Comment by Siedlerchr** [comment]
except for the submodule stuff it looks good from my side! Thanks for your contribution and the hopefully onoging interest in JabRef!

**Comment by heyitsdross** [comment]
Regarding the submodules: 
1. The steps above don't seem to have removed? them from the changelist in my branch. What is the desired change I'm trying to enact here, a removal? And is there more in-depth reading I can use to get a better understanding of what I'm doing here?
2. Are the submodules not supposed to be checked in? I have myself little experience with them, execpt for one project at work, where git simply kept a commit hash in place of the files, which one could pull via `git submodule update` or similar.

**Comment by subhramit** [comment]
> What is the desired change I'm trying to enact here, a removal?

Right now, the desired change would be that you sync the submodules in your PR with the ones in `Jabref:main` right now.
> 2. Are the submodules not supposed to be checked in? I have myself little experience with them, execpt for one project at work, where git simply kept a commit hash in place of the files, which one could pull via `git submodule update` or similar.

Yes, `git submodule update` does update the submodules. But for committing them, you will have to go to `.gitmodules` and comment ou the line that sets them to `ignored`. After committing, revert the `.gitmodules` configuration, and push - everything should be synced then (that is one way to do it). I think @Siedlerchr might have a better answer to this.

**Comment by heyitsdross** [comment]
I very much appreciate the help, but I'm totally lost. I'll need to work experiment with my own submodules on my own time to understand how to sync them properly

**Comment by Siedlerchr** [comment]
The problem with submodules is often when you create a new branch with IntelliJ, it will create a new branch for the submodules as well. If you now fetch the latest updates in IntelliJ it will update the submodules as well and ignore the "pinned" commit from main. In the commit window, it will now list the changes to the submodules as well. One has to explicit uncheck them

TL;DR: reset submodules
edit .gitmodules
comment out ignore = all
`#ignore = all`

cd buildres/abbrv.jabref.org 
git checkout [234cce8](https://github.com/JabRef/abbrv.jabref.org/tree/234cce883adf9012c2831990d883b198d1d6a939)

same for csl styles:
cd src/main/resources/csl-styles
git checkout [4e0902d](https://github.com/citation-style-language/styles/tree/4e0902d4885da1e910e5898a644778bbc6b6c0d6)

use git gui to commit **submodule changes only.** 
Afterwards revert the changes in .gitmodules

File: .gitmodules [hunk 5]
```diff
@@ -1,15 +1,15 @@
 [submodule "abbrv.jabref.org"]
   path = buildres/abbrv.jabref.org
   url = https://github.com/JabRef/abbrv.jabref.org.git
-  ignore = all
+# ignore = all
```
Comment (Siedlerchr) [inline .gitmodules:4]: see my latest comment for a TL;DR (git submodules are a pain in the a**)

**Comment by heyitsdross** [comment]
Thank you two very much for the help! I believe the submodule changes are synced again.

**Review by Siedlerchr — APPROVED** [review]

**Review by subhramit — APPROVED** [review]

**Comment by subhramit** [comment]
@heyitsdross by the way, if you use "Closes" in your PR description, it is treated as a keyword and the issue is linked automatically.

**Comment by heyitsdross** [comment]
So that's how that works! Very nice :)

**Comment by Siedlerchr** [comment]
@heyitsdross Thanks again for your PR! Keep up the good work :)
 Fixes and closes and some other keywords are possible  https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests
